### PR TITLE
Add depth limiter to `isConstTypeVariable` function

### LINKED ISSUE #54610: `isConstTypeVariable` stack overflow
# Bug Report

### 🔎 Search Terms

`isConstTypeVariable`

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions 5.0 and 5.1.3

I suspect it's related to the changes in #53341 (cc @ahejlsberg)

### ⏯ Playground Link

I trimmed our product code to obtain a minimal repro. This code doesn't make sense and doesn't preserve the semantic of what the original code does (`B` was originally an helper retrieving an union of all possible paths for an object), but it's much shorter and still reproduces the stack overflow.

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggPAFQHxQLxQQbgLACg+iRQBCiK6CUEAHsBAHYAmAzlAIb0gDaAunlAKgB+KPQgA3CACd+ggFywyVWgxZQAlvQBm0qAGsIIWQJEGQyuk1ZmA9loxQAZFGbApmgObHBJkoi5mPEjePgpikjK4oaIS0jj4uFoArvQAxsDqNvRQWjY2ZAAUAJRQAN7eAAYAJKVsAL4V7KykyPF1eEA)

### 💻 Code

```ts
type A<T> = T;

type B<T> = T extends any[]
    ? never
    : A<T> extends infer key
    ? key extends keyof T
        ? B<T[key]>
        : never
    : never;

function foo<T>() {
    `${a}` as B<T>;
}
```

### 🙁 Actual behavior

Stack overflow

### 🙂 Expected behavior

Shouldn't stack overflow (various errors should be reported, but not a stack overflow).

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #54610.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Review by RyanCavanaugh — APPROVED** [review]

**Comment by ahejlsberg** [comment]
Latest commit removes the first fix and instead adds a depth limiter to `isConstTypeVariable`. After more careful analysis I realized that was the source of the infinite recursion.

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot perf test faster
@typescript-bot test top100

**Comment by jakebailey** [comment]
@typescript-bot run dt

But this time don't time out on npm

**Comment by ahejlsberg** [comment]
Ok, I think we finally have a clean test and perf run. This one is ready to merge.

**Review by sandersn — APPROVED** [review]

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -13538,14 +13538,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
-            type.flags & TypeFlags.Union && some((type as UnionType).types, isConstTypeVariable) ||
-            type.flags & TypeFlags.IndexedAccess && isConstTypeVariable((type as IndexedAccessType).objectType) ||
-            type.flags & TypeFlags.Conditional && isConstTypeVariable(getConstraintOfConditionalType(type as ConditionalType)) ||
-            type.flags & TypeFlags.Substitution && isConstTypeVariable((type as SubstitutionType).baseType) ||
-            isGenericTupleType(type) && findIndex(getElementTypes(type), (t, i) => !!(type.target.elementFlags[i] & ElementFlags.Variadic) && isConstTypeVariable(t)) >= 0));
+            type.flags & TypeFlags.Union && some((type as UnionType).types, t => isConstTypeVariable(t, depth)) ||
+            type.flags & TypeFlags.IndexedAccess && isConstTypeVariable((type as IndexedAccessType).objectType, depth + 1) ||
+            type.flags & TypeFlags.Conditional && isConstTypeVariable(getConstraintOfConditionalType(type as ConditionalType), depth + 1) ||
+            type.flags & TypeFlags.Substitution && isConstTypeVariable((type as SubstitutionType).baseType, depth) ||
```
[the remark this answers is not in the document]
Comment (ahejlsberg) [reply src/compiler/checker.ts:13547]: Exactly!

**Pull request merged**